libraries/AP_HAL_Linux/Storage.cpp

```diff
@@ -20,24 +20,103 @@ using namespace Linux;
 
 // name the storage file after the sketch so you can use the same board
 // card for ArduCopter and ArduPlane
-#if APM_BUILD_TYPE(APM_BUILD_Replay)
-#define STORAGE_DIR "."
-#else
-#define STORAGE_DIR HAL_BOARD_STORAGE_DIRECTORY
-#endif
-
-#define STORAGE_FILE STORAGE_DIR "/" SKETCHNAME ".stg"
+#define STORAGE_FILE SKETCHNAME ".stg"
 
 extern const AP_HAL::HAL& hal;
 
+static inline int is_dir(const char *path)
+{
+    struct stat st;
+
+    if (stat(path, &st) < 0) {
+        return -errno;
+    }
+
+    return S_ISDIR(st.st_mode);
+}
+
+static int mkdir_p(const char *path, int len, mode_t mode)
+{
+    char *start, *end;
+
+    start = strndupa(path, len);
+    end = start + len;
+
+    /*
+     * scan backwards, replacing '/' with '\0' while the component doesn't
+     * exist
+     */
+    for (;;) {
+        int r = is_dir(start);
+        if (r > 0) {
+            end += strlen(end);
+
+            if (end == start + len) {
+                return 0;
+            }
+
+            /* end != start, since it would be caught on the first
+             * iteration */
+            *end = '/';
+            break;
+        } else if (r == 0) {
+            return -ENOTDIR;
+        }
+
+        if (end == start) {
+            break;
+        }
+
+        *end = '\0';
+
+        /* Find the next component, backwards, discarding extra '/'*/
+        while (end > start && *end != '/') {
+            end--;
+        }
+
+        while (end > start && *(end - 1) == '/') {
+            end--;
+        }
+    }
+
+    while (end < start + len) {
+        if (mkdir(start, mode) < 0 && errno != EEXIST) {
+            return -errno;
+        }
+
+        end += strlen(end);
+        *end = '/';
+    }
+
+    return 0;
+}
+
 void Storage::_storage_create(void)
 {
-    mkdir(STORAGE_DIR, 0777);
-    unlink(STORAGE_FILE);
-    int fd = open(STORAGE_FILE, O_RDWR|O_CREAT|O_CLOEXEC, 0666);
+    const char *dpath = HAL_BOARD_STORAGE_DIRECTORY, *p;
+    int dfd = -1;
+
+    p = hal.util->get_custom_storage_directory();
+    if (p) {
+        dpath = p;
+    }
+
+    mkdir_p(dpath, strlen(dpath), 0777);
+    dfd = open(dpath, O_RDWR|O_CLOEXEC);
```

Comment:
Did you test `open()` to work for a directory the way you use it? 

Try following in `test.cpp` in any linux setup:
```
#define HAL_BOARD_STORAGE_DIRECTORY "/tmp/ardupilot"

static inline int is_dir(const char *path)
{
...
}
static int mkdir_p(const char *path, int len, mode_t mode)
{
...
}
int main() {
  const char *dpath = HAL_BOARD_STORAGE_DIRECTORY;
  int dfd = -1;
  mkdir_p(dpath, strlen(dpath), 0777);
  dfd = open(dpath, O_RDWR|O_CLOEXEC);
  printf("Error opening storage directory: %s - %i\n", dpath, dfd);
}


$ gcc test.cpp
$ ./a.out 
Error opening storage directory: /tmp/ardupilot - -1
$ file /tmp/ardupilot/
/tmp/ardupilot/: directory
```

Reply:
Yep, I had this problem when testing last week and fixed it here... It's the reason why I didn't push this yet. Will fix it soon.